---
title: "How to format the ARB-exported FreshTrain for use with TaxAss"
author: "Robin Rohwer"
date: "last updated May 28, 2020"
output: 
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{css, echo=FALSE}
pre code, pre, code {
  white-space: pre !important;
  overflow-x: scroll !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
```

# Get Started  

## Overview of steps  

* do a first pass of reformatting FreshTrain  
  * export all full-length 16S from ARB  
  * basic text-based reformatting of taxonomy and fasta files  
  * fix duplicated taxon names and standardize unnamed nomenclature
  * check that only and all expected FreshTrain names are included  
* do a first pass of reformatting Silva
  * basic text-based reformatting of fasta file
  * fix duplicated taxon names and standardize unnamed nomenclature
* identify coarse-level nomenclature disagreements between the databases
  * classify FreshTrain with SILVA  
  * compare coarse-level disagreements
* change FreshTrain coarse-level nomenclature to match Silva  
  * create a custom python script to reformat FreshTrain names
* manually check for non-monophyletic sequences
  * manually adjust them
  
<br>

## Set Up Working Directory

Move all needed scripts & files into a new working directory folder:

location    | filename  
------------|-------------------------------
arb export  | version.txt  
arb export  | version.fasta  
mothur silva download | version.tax  
mothur silva download | version.align
arb-scripts    | remove_gg_taxon-level_prefixes.sh  
arb-scripts    | replace_first_semicolon_with_tab.sh  
arb-scripts    | check_included_lineages.R
tax-scripts    | reformat_fasta.R
tax-scripts    | reformat_taxonomy_nomenclature.R


<br>

# First-Pass FreshTrain

## ARB export  

The FreshTrain comes from an ARB database of 16S clone libraries from freshwater systems. 
All of the clones were aligned in ARB alongside the Greengenes database to determine the 
phylogeny of abundant freshwater taxa introduced in [Newton *et al.* 2011](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3063352/). 

**Format of raw exports looks like:**  
```
$ head -n 5 FreshTrain_version.fasta
>CkeE0000
......................A-GAGUUU-GAU-CAUGGCUCAG-GAU-GAAC-GCUAGCGGGAGGCUU-AAUACAUGCAAGUCGAGGG--AC-AG-CAC------AGGUA-GCAA-UA--CUG-----------G-GUGGU--AACCG-G-C--AA-ACGGG-UGCGG-AACACGUACACAACCUUC-CUUUAAG-UGGGGGAUAGCCCAG-A-GAAA-UUUG-GA-UUAAUACCCCGU-AACA-CAGU--G-GAGCG-GCAUCGC--UCU-AUUGUUAUA--GA-------------UACG------------UCGC-UUAAUGAUGGGUGUGCGUCUGAUUAGAUAG-UUGGCGAGGUAACGGCUCACCAA-GUCUACGAUCAGUAG-CUGAUGUGAGA-GCAUGAUCAGCCACACGGGCA-CUGAGACACG-GGCCCGA-CUCC-UACGGGAGGC-AGCAGUAAGGAAUAUU-G-GUUCAAUGGACGCAAGUCUGAACCAGC-CAU-GCC-GCGUGAAGGAUUAA-GGUCCUC-UG--GA-UUGUAAA-CUUCUUUU-AUUU-GG-GAC-GA-AAAAGGGUCU----UUCU-----A-G-AC-CAA-CUGACGGUACC-AGA-UG-AAUAAGCACCGG-CU-AACUCCGUGCCA-GCAGCC-GC-GGUAAUAC-GGAG-GGUGC-AAGCGUUAUCC-GGAUUCA-CUGGGU-UUAAA-GGGUGCG-UAGGCGGGUAGGU-AA-GUCA-G-UGGUGAAAUCCUG-GAG-C-UCAACUU-CA-G-AAC-UG-CCAUU-GAUACU-AUUU-AUCUUG-AAUAUUGUGGAGGUAAGC-GGAAUA-UGUCAU-GUA-GCG-GUGAAA-UGC-UU-AGAU-AUG-ACAUAG-AAC-A-CCUAUUGC-GAA-GGCAGCUUACUA-CGCA-UAUAUUGACG-CUGAGG-CAC-GAAAGCGU-GG-GGAUCAAACA-GGAUUA-GA-UACCCU-GG-UA-GUC-CACGCC-CUAAAC-UAU-GGA-UACUCGACAU-UUGU-----GAUACACA------AU-AAG-UGUCUGA--GC-GAA-AG-CA-UUAA-GUAU-CC-CACCU-GGG-AAGUAC-GUCC--GCAA-G-GAU-GAAACUC-AAA-GGAAU-U-G-GCGGG-G-GUUCC-GCACA-AGC-G-GUGG-AGCAU-GU-GGU-UUAA-UU-CG-AUGAUACG-CGAG-GAA-CC-UUA-CCUGGGC--UAGAAU-GC-U-G-AGAG----ACCGU-GG-GUGAAA-GCC-C-AUU-U--UG---UAGC--AA-UA----CA-CUGUC-AGUA--AGGU-GC-UGCAUGGCUGUCGUCAGCU-CGUUGC-CGUGAGGUGUUGGG-UUAAGU-CCCGCAACGAGCGCAACCCCUA-UCAUU-AG-UUGC-CAA-CAG-GUCAAG----CCGGG-AACUCUAAUGAAACUG-CCGCC--GUAAGGC-GUGAGGA-AGGAGGGG-AU-GAUGUCAAGUC-AUCAUGGC-CUUUAUGCCCA-GGGC-UACACACGUGCUACAAUGGA-C-AG-GA-CAAAGGGCUGCCACCU-AGUGA-UAGG-GCGCUAA-UCCCAA-AAA-C-CU-G-UUCUCAGUU-CAGAUUGGAGUCUGCAACUC-GACUCCAUGAA-GCUGGAAUCGC-U-AGUAAU-CGU-AUA-UCAGCAAU-GAUA-CGGUG-AAUACGUU-CCCGGA-CCUUGCACACACCGCCC-GUCAAGCCA-UGGGA--GCCGGGU-GUACCU-AAAG-UCGGU-------A-C-----------GCAG---GA-----GCUGCC-UA-GG-GU-AA--AAUC-GG-UGACU-GGGG-CU---AAGUC-GUAA-C--AAGG-UA-GCC-GUACC-GGA-A-GGUG--CGGC--UG-GAAUACCUCCU..................................................................................
>CbfYyyyy
......................A-GAGUUU-GAU-CAUGGCUCAG-AUU-GAAC-GCUGGCGGCAUGCUU-UACACAUGCAAGUCGAACG--GC-AG-CGC--------GGG-GCAA-CC--U-------------G-GCGGC--GAGUG-G-C--GA-ACGGG-UGAGU-AAUACAUCGGA-ACGUAC-CCAGAAG-UGGGGGAUAACGUAG-C-GAAA-GUUA-CG-CUAAUACCGCAU-ACG------------UUC-UACGGAA-----------GAAA--GUGGGGGACC-----UUCG-----GG-CCUCAUGC-UUUUGGAGCGGCCGAUGUCUGAUUAGCUAG-UUGGUGAGGUAAAGGCUCACCAA-GGCGACGAUCAGUAG-CUGGUCUGAGA-GGACGACCAGCCACACUGGGA-CUGAGACACG-GCCCAGA-CUCC-UACGGGAGGC-AGCAGUGGGGAAUUUU-G-GA-CAAUGGGCGCAAGCCUGAUCCAGC-AAU-GCC-GCGUGAGUGAAGAA-GGCC-UU-CG--GG-UUGUAAA-GCUCUUUU-GUCA-GG-GAA-GA-A-ACGGCUUAGG-UUAAUAC-CUU-C-GG-CUA-AUGACGGUACC-UGA-AG-AAUAAGCACCGG-CU-AACUACGUGCCA-GCAGCC-GC-GGUAAUAC-GUAG-GGUGC-AAGCGUUAAUC-GGAAUUA-CUGGGC-GUAAA-GCGUGCG-CAGGCGGUUUUGU-AA-GUCU-G-ACGUGAAAUCCCC-GGG-C-UCAACCU-GG-G-AAU-UG-CGUUG-GAGACU-GCAA-GGCUGG-AGUCUGGCAGAGGGGGGU-AGAAUU-CCACGU-GUA-GCA-GUGAAA-UGC-GU-AGAG-AUG-UGGAGG-AAC-A-CCGAUGGC-GAA-GGCAGCCCCCUG-GGUC-AAGACUGACG-CUCAUG-CAC-GAAAGCGU-GG-GGAGCAAACA-GGAUUA-GA-UACCCU-GG-UA-GUC-CACGCC-CUAAAC-GAU-GUC-UACUAGUUGU-CGGGUC----UUAAUU---G-A-CU-UGG-UAACGCA--GC-UAA-CG-CG-UGAA-GUAG-AC-CGCCU-GGG-GAGUAC-GGUC--GCAA-G-AUU-AAAACUC-AAA-GGAAU-U-G-ACGGG-G-A-CCC-GCACA-AGC-G-GUGG-AUGAU-GU-GGA-UUAA-UU-CG-AUGCAACG-CGAA-AAA-CC-UUA-CCUACCC--UUGACA-UG---U-CAGG----AACCU-CC-GAGAGA-UUG-G-AGG-G--UG--CCCGA--AA-GGG---AG-CCUGA-ACAC--AGGU-GC-UGCAUGGCUGUCGUCAGCU-CG-UGU-CGUGAGAUGUUGGG-UUAAGU-CCCGCAACGAGCGCAACCCUUG-UCAUU-AG-UGUC-UA--C----GCAA-------GAG-CACUCUAAUGAGACUG-CCGGU-GACAAACC-G-GAGGA-AGGUGGGG-AU-GACGUCAAGUC-CUCAUGGC-CCUUAUGGGUA-GGGC-UUCACACGUCAUACAAUGGU-G-CA-UA-CAGAGGGCCGCCAACC-CGCGA-GGGG-GAGCUAA-UCCCAG-AAA-G-UG-C-AUCGUAGUC-CGGAUCGCAGUCUGCAACUC-GACUGCGUGAA-GUUGGAACCGC-U-AGUAAU-CGC-GGA-UCAGCAU--GUCG-CGGUG-AAUACGUU-CCCGGG-UCUUGUACACACCGCCC-GUCACACCA-UGGGA--GCGGGUU--UACCA-GAAG-UAGGUAG-CUUAA-CC----------GCAA--GGAGG-GCGCUUAC-CA-CG-GU-AG--GAUU-CG-UGACU-GGGG-UG---AAGUC-GUAA-C--AAGG-UA-GCC-GUAUC-GGA-A-GGUG--CGGC--UG-GAUCACCUCCUU.................................................................................
>CslBog00

```
**Note that:**  

* The fasta file contains U's instead of T's
* The fasta file is aligned (contains dots `.` and dashes `-`)

```
$ head -n 5 FreshTrain_version_taxonomy.txt
CkeE0000,k__Bacteria;p__Bacteroidetes;c__Sphingobacteriia;o__Sphingobacteriales;bacI;bacI-A;bacI-A2;
CbfYyyyy,k__Bacteria;p__Proteobacteria;c__Betaproteobacteria;o__Burkholderiales;betVII;betVII-A;betVII-A1;
CslBog00,k__Bacteria;p__Proteobacteria;c__Betaproteobacteria;o__Burkholderiales;betVII;
CslBog02,k__Bacteria;p__Proteobacteria;c__Betaproteobacteria;o__Burkholderiales;betII;Pnec;PnecC;
ShnBacte,k__Bacteria;p__Bacteroidetes;c__Flavobacteriia;o__Flavobacteriales;bacII;
```

**Note that:**  

* The taxonomy file has a comma delimiter btwn seqID and taxonomy names  
* Some references with FreshTrain names are blank at finer taxon resolutions    

In arb there are also references that don't have good representation or specific FreshTrain names, and these is a field that marks each sequence as part of the FreshTrain or not. It's important that only the FreshTrain sequences are exported, so you can spot check for unclassified lineages here. We will also check for these more carefully later.  


<br>


## fasta formatting

#### Replace `U` with `T` and unalign (R)  

```{bash, eval=F}
Rscript reformat_fasta.R FreshTrain_version.fasta FT_unaligned.fasta format.ids replace.U
```
Argument  | Description  
----------|---------------
`reformat_fasta.R` | Name of script.
`FreshTrain_version_taxonomy.txt` | input mothur-formatted taxonomy file
`FreshTrain_unaligned.tax` | output mothur-formatted taxonomy file with no no U's, dots, or dashes
`format.ids` | this doesn't do anything because there's no blanks in the seqIDs anyway. (If there were, it would delete them and everything after)
`replace.U` | this tells it to find all the `U`s or `u`s and replace them with `T`  

Now it looks like:  
```
$ head -n 5 FT_unaligned.fasta
>ClonLD29
AACGAACGCTGGCGGCGTGGATAAGACATGCAAGTCGAACGGGATTTTTTTCTGTAGCAATACAGAAAAAAGTCTAGTGGCGAACGGGTGCGTAACACGTGAGCAACCTACCGTGAAGTGGGGGATAGCTCGCCGAAAGGCGAATTAATACCGCATGTGGCCAAGGAAGACATCTTCCTGACGCCAAAGCCGGGGTAACCTGGCGCTTCTCGAGGGGCTCGCGGCCTATCAGCTTGTTGGTGAGGTAACGGCTCACCAAGGCTATGACGGGTAGCTGGTCTGAGAGGACGACCAGCCACACTGGAACTGAGACACGGTCCAGACACCTACGGGTGGCAGCAGTCGAGAATTTTTCACAATGGGGGAAACCCTGATGGAGCGACGCCGCGTGGAGGATGAAGGTTTTCGGATCGTAAACTCCTGTCATTAGAGAACAAGGCACATGGTTTAACTGACTGTGTGTTGATAGTATCTGAAGAGGAAGGGACGGCTAACTCTGTGCCAGCAGCCGCGGTAATACAGAGGTCTCAAGCGTTGTTCGGATTCATTGGGCGTAAAGGGTGCGTAGGTGGCGAGGTAAGTCGGATGTGAAATCTCCAAGCTCAACTTGGAAACTGCATTCGATACTGCTTTGCTAGAGGATTGTAGAGGGCATTGGAATTCACGGTGTAGCAGTGAAATGCGTAGATATCGTGAGGAAGACCAGTGGCGAAGGCGAATGCCTGGGCAATTCCTGACACTGAGGCACGAAGGCCAGGGGAGCAAACGGGATTAGATACCCCGGTAGTCCTGGCAGTAAACGGTGCACGTTTGGTGTGGGAGGATTCGACCCCTTCCGTGCCGGAGCTAACGCGTTAAACGTGCCGCCTGGGGAGTACGGTCGCAAGATTAAAACTCAAAGAAATTGACGGGGACCCGCACAAGCGGTGGAGTATGTGGCTTAATTCGATGCAACGCGAAGAACCTTACCTGGCCTTGACATGCATCTCTAATCCGGTGAAAGCCGGTGAGTGTAGCAATACACAATTTGCACAGGTGCTGCATGGCTGTCGTCAGCTCGTGTCGTGAGATGTTGGGTTAAGTCCCGCAACGAGCGCAACCCCTGTGAACTGTTGCCATGGAGACGTGAGTCTCCGGCACTCTGTTCAGACTGCCCTGTTCAACGGGGAGGAAGGTGGGGACGACGTCAAGTCAGTATGGCCCTTACGGCCAGGGCTGCACACGTACTACAATGCCCGACACAGAAGGATCCGAGACCGCAAGGTGGAGGAAATCACCAAAGTCGGGCCCAGTTCGGATTGGAGGCTGCAACTCGCCTCCATGAAGCCGGAATCGCTAGTAATGGCGCATCAGCTACGGCGCCGTGAATACGTTCCCGGGTCTTGTACACACCGCCCGTCACATCATGGGAGTCGTTTGTACCCGAAGTGCGCGCGCCAACCGCAAGGAGGCAGCGCCCTAAGGTATGAGCGGTAACTGGGATG
>FlvAquat
NTACGATGAAGAGTTTGATCCTGGCTCAGGATGAACGCTAGCGGCAGGCTTAACACATGCAAGTCGAGGGGTATATGTCTTCGGATATAGAGACCGGCGCACGGGTGCGTAACGCGTATGCAATCTACCTTTTACAGGGGAATAGCCCAGAGAAATTTGGATTAATGCCCCATAGTATTATTAAATGGCATCATTTTATAATTAAAGTCACAACGGTAAAAGATGAGCATGCGTCCCATTAGTTAGTTGGTAAGGTAACGGCTTACCAAGACAATGATGGGNAGGGGTCCTGAGAGGGAGATCCCCCACACTGGTACTGAGACACGGACCAGACTCCTACGGGAGGCAGCAGTGAGGAATATTGGACAATGGGCGCAAGCCTGATCCAGCCATGCCGCGTGCAGGAAGACGGTCCTATGGATTGTAAACTGCTTTTATACAGGAAGAAACACTCCGACGTGTCGGAGCTTGACGGTACTGTAGGAATAAGGATCGGCTAACTCCGTGCCAGCAGCCGCGGTAATACGGAGGATCCAAGCGTTATCCGGAATCATTGGGTTTAAAGGGTCCGTAGGCGGTTTAATAAGTCAGTGGTGAAATCTGGTCGCTCAACGATCAAACGGCCATTGATACTGTTAGACTTGAATTATAAGGAAGTAACTAGAATATGTAGTGTAGCGGTGAAATGCTTAGATATTACATGGAATACCAATTGCGAAGGCAGGTTACTACTTATGGATTGACGCTGATGGACGAAAGCGTGGGGAGCGAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGGATACTAGCTGTTGGGAGCAATCTCAGTGGCTAAGCGAAAGTGATAAGTATCCCACCTGGGGAGTACGTTCGCAAGAATGAAACTCAAAGGAATTGACGGGGGCCCGCACAAGCGGTGGAGCATGTGGTTTAATTCGATGATACGCGAGGAACCTTACCAAGGCTTAAATGTAGATTGACAGGACTGGAAACAGTTTTTTCTTCGGACAATTTACAAGGTGCTGCATGGTTGTCGTCAGCTCGTGCCGTGAGGTGTCAGGTTAAGTCCTATAACGAGCGCAACCCCTGTTGTTAGTTGCCAGCGAGTCATGTCGGGAACTCTAACGAGACTGCCAGTGCAAACTGTGAGGAAGGTGGGGATGACGTCAAATCATCACGGCCCTTACGCCTTGGGCTACACACGTGCTACAATGGACGGTACAGAGAGCAGCCACTGCGCAAGCAGGAGCGAATCTACAAAACCGTTCTCAGTTCGGATCGGAGTCTGCAACTCGACTCCGTGAAGCTGGAATCGCTAGTAATCGGATATCAGCCATGATCCGGTGAATACGTTCCCGGGCCTTGTACACACCGCCCGTCAAGCCATGGAAGCTGGGGGTACCTGAAGTCGGTGACCGCAAGGAGCTGCCTAGGGTAAAACTGGTAACTGGGGCTAAGTCGTAACAAGGTAGCCGTACCGGAAGG
>B61FlSp2
```

<br>

## taxonomy formatting  

#### Make file semicolon delimited (bash)  

```{bash, eval=F}
sed 's/,/\;/' <FreshTrain_version_taxonomy.txt >FT_semicol.tax
```

Now it looks like:
```
$ head -n 5 FT_semicol.tax
ClonLD29;k__Bacteria;p__Verrucomicrobia;c__Spartobacteria;o__Chthoniobacterales;verI-A;
FlvAquat;k__Bacteria;p__Bacteroidetes;c__Flavobacteriia;o__Flavobacteriales;bacII;bacII-A;
B61FlSp2;k__Bacteria;p__Bacteroidetes;c__Cytophagia;o__Cytophagales;bacIII;bacIII-A;
St0rFro8;k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Rhodobacterales;alfVI;
St0rFr10;k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;alfIV;alfIV-A;Novo-A2;
```

<br>

#### Remove k\_\_, p\_\_, c\_\_, o\_\_ (bash)  

```{bash, eval=F}
./remove_gg_taxon-level_prefixes.sh FT_semicol.tax FT_semicol_noprefix.tax
```
Argument  | Description  
----------|---------------  
`./remove_gg_taxon-level_prefixes.sh` | name of script  
`FT_semicol.tax`  | The input taxonomy file, semicolon delimited      
`FT_semicol_noprefix.tax` | The output taxonomy file, same but without `k__` `p__` `c__` or `o__` in front of the coarse-level names. This makes it compatible with silva, because only greengenes uses those prefixes.    

Now it looks like:  
```
$ head -n 5 FT_semicol_noprefix.tax
ClonLD29;Bacteria;Verrucomicrobia;Spartobacteria;Chthoniobacterales;verI-A;
FlvAquat;Bacteria;Bacteroidetes;Flavobacteriia;Flavobacteriales;bacII;bacII-A;
B61FlSp2;Bacteria;Bacteroidetes;Cytophagia;Cytophagales;bacIII;bacIII-A;
St0rFro8;Bacteria;Proteobacteria;Alphaproteobacteria;Rhodobacterales;alfVI;
St0rFr10;Bacteria;Proteobacteria;Alphaproteobacteria;Sphingomonadales;alfIV;alfIV-A;Novo-A2;
```

<br>

#### Make first column tab delimited for mothur-compatibility (bash)  

```{bash, eval=F}
./replace_first_semicolon_with_tab.sh FT_semicol_noprefix.tax FT_semicol_noprefix_mothur.tax
```
Argument  | Description  
----------|---------------
`FT_semicol_noprefix.tax` | The input taxonomy file that is semicolon delimited    
`FT_semicol_noprefix_mothur.tax` | The output taxonomy file that has semicolons between taxon-levels but a tab between the sequence ID and the taxa-names. This is the mothur format for taxonomy files.  

Now it looks like:  
```
$ head -n 5 FT_semicol_noprefix_mothur.tax
ClonLD29	Bacteria;Verrucomicrobia;Spartobacteria;Chthoniobacterales;verI-A;
FlvAquat	Bacteria;Bacteroidetes;Flavobacteriia;Flavobacteriales;bacII;bacII-A;
B61FlSp2	Bacteria;Bacteroidetes;Cytophagia;Cytophagales;bacIII;bacIII-A;
St0rFro8	Bacteria;Proteobacteria;Alphaproteobacteria;Rhodobacterales;alfVI;
St0rFr10	Bacteria;Proteobacteria;Alphaproteobacteria;Sphingomonadales;alfIV;alfIV-A;Novo-A2;
```

<br>


#### Fill missing names and fix any duplicate names in the FreshTrain (R)  

```{bach, eval=F}
$ Rscript reformat_taxonomy_nomenclature.R FT_semicol_noprefix_mothur.tax FT_semicol_noprefix_mothur_unnamed.tax FreshTrain
```  
Argument  | Description  
----------|---------------
`reformat_taxonomy_nomenclature.R` | Name of script.  
`FT_semicol_noprefix_mothur-fmt.tax` | The input taxonomy file, mothur-formatted.  
`FT_semicol_noprefix_mothur-fmt_unnamed.tax` | The output taxonomy file, also mothur-formatted. All missing names (blanks or placeholders indicating there is no name at that level) have been changed to `unnamed.NearestCoarserName`. Any duplicate names have had suffixes added to make them unique.  
`FreshTrain` | This flag is necessary because it tells the script to use `.l` `.c` `.t` instead of `.F` `.G` `.S` as level-suffixes when it makes duplicated names unique.  

The top-of-script descriptions of what this does, along with examples, is very detailed.  
  
The terminal output for this is extensive and informative- read it carefully for mistakes where names changed that shouldn't have.   
It tells you which names are changed, for example:  

  * alfI-A is a lineage and clade, so it is changed to alfI-A.l and alf_I-A.c .  
  * Similarly, it lets you know that 434 blanks were changed to say "unnamed".  
<br>

It is really important that blanks are filled with placeholder "unnamed..." becuase the RDP classifier takes a computational shortcut when calculating bootstrap p-values. For higher taxonomy levels it simply sums the lower levels.  If the lower level is totally blank there's nothing to sum up, so the highest level ends up being less than 100%.  

Now it looks like this:
```
$ head -n 5 FT_semicol_noprefix_mothur_unnamed.tax
ClonLD29	Bacteria;Verrucomicrobia;Spartobacteria;Chthoniobacterales;verI-A;unnamed.verI-A;unnamed.verI-A;
FlvAquat	Bacteria;Bacteroidetes;Flavobacteriia;Flavobacteriales;bacII;bacII-A;unnamed.bacII-A;
B61FlSp2	Bacteria;Bacteroidetes;Cytophagia;Cytophagales;bacIII;bacIII-A;unnamed.bacIII-A;
St0rFro8	Bacteria;Proteobacteria;Alphaproteobacteria;Rhodobacterales;alfVI;unnamed.alfVI;unnamed.alfVI;
St0rFr10	Bacteria;Proteobacteria;Alphaproteobacteria;Sphingomonadales;alfIV;alfIV-A;Novo-A2;
```

<br>

## Check for Non-FreshTrain (R)  

The full set of freshwater sequences in arb includes all of the clones, including ones that are not part of the well-defined freshwater taxonomy. These non-FreshTrain sequences cannot be present for TaxAss because only named reference sequences should recruit reads into the FreshTrain classification.  

Within Arb, sequences are identified as part of the FreshTrain or not, and when data is exported from Arb you should select to only export the FW sequences. Here we check to make sure that only FreshTrain sequences are included (because it's really bad if they are!):    

```{bash, eval=F}
$ Rscript check_included_lineages.R FT_semicol_noprefix_mothur_unnamed.tax ./
```
Argument  | Description  
----------|---------------
`check_included_lineages.R` | Name of the script  
`FT_semicol_noprefix_mothur_unnamed.tax` | Name of the taxonomy file. mothur formatted, and blanks must have been filled with "unnamed..." already.  
`./`   | The path to the folder where generated files should go.  

It there are sequences where the lineage is unnamed, these are flagged as sequences to definitely remove. This is exported in the file `STOP-remove_these_from_FreshTrain_export.csv` in csv format for easy opening in excel:  
```
$ head -n 5 STOP-remove_these_from_FreshTrain_export.csv
seqid,kingdom,phylum,class,order,family,genus,species
SphClo28,Bacteria,Bacteroidetes,unnamed.Bacteroidetes,unnamed.Bacteroidetes,unnamed.Bacteroidetes,unnamed.Bacteroidetes,unnamed.Bacteroidetes
SphClo38,Bacteria,Bacteroidetes,unnamed.Bacteroidetes,unnamed.Bacteroidetes,unnamed.Bacteroidetes,unnamed.Bacteroidetes,unnamed.Bacteroidetes
MbGubG03,Bacteria,Bacteroidetes,unnamed.Bacteroidetes,unnamed.Bacteroidetes,unnamed.Bacteroidetes,unnamed.Bacteroidetes,unnamed.Bacteroidetes
MbHubH00,Bacteria,Bacteroidetes,unnamed.Bacteroidetes,unnamed.Bacteroidetes,unnamed.Bacteroidetes,unnamed.Bacteroidetes,unnamed.Bacteroidetes
```

This script always creates 3 other files, that list the unique names at each FreshTrain level (lineage, clade, tribe). Since it is easy for the inclusion to be toggled in arb, these should be checked to make sure that:  

* no non-FreshTrain names appear  
* all expected FreshTrain names appear  

The 3 files are called:  

* `Check_this_list_of_unique_lineages_included_in_FreshTrain_export.csv`  
* `Check_this_list_of_unique_clades_included_in_FreshTrain_export.csv`
* `Check_this_list_of_unique_tribes_included_in_FreshTrain_export.csv`  

These are all csv taxonomy files that look like:  
```
$ head -n 5 Check_this_list_of_unique_lineages_included_in_FreshTrain_export.csv
kingdom,phylum,class,order,family
Bacteria,Actinobacteria.P,Acidimicrobiia,Acidimicrobiales,acIV
Bacteria,Actinobacteria.P,Acidimicrobiia,Acidimicrobiales,acV
Bacteria,Actinobacteria.P,Actinobacteria.C,Corynebacteriales,acTH2
Bacteria,Actinobacteria.P,Actinobacteria.C,Frankiales,acI
```

<br>
<br>


# Silva First-Pass

## Download  

#### Download Silva from mothur website  

https://mothur.org/wiki/silva_reference_files/  
Choose the "Full-length sequences and taxonomy references" file.  

**The mothur-formatted files look like:**  
```
$ head -n 5 silva.nr_v138.align
>CP026050.S49M2367	99.47	Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Yersiniaceae;Serratia;
...................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................A-T-T--G-A--A-C-G-C--T-G-G-C--G-G--C-A-G-G----------C----T-T--AA-C--AC-A----T-G--C----A-A-G--T-C--GA-G-CG-----------G-TAG--C-A--C-------------------------------AG-G-G-G-A-G---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------CTT-G----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------C--T--C-C-C-C--G------------------G-G-T----G--AC--G-------AG-C-G-G-C-GG-A--C-------------G-G-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------G--T--G-A--G-T-A---AT-G-T-C--T-G-GG---A-A--A-CT-G--C-C-TGA--TG-G-----------------(deleted rest of line)...
>JQ864435.O10Acti2	93.57	Bacteria;Proteobacteria;Alphaproteobacteria;Rhodobacterales;Rhodobacteraceae;Oceanicella;
...................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................A-A-C--G-A--A-C-G-C--T-G-G-C--G-G--C-A-G-G----------C----C-T--AA-C--AC-A----T-G--C----A-A-G--T-C--GA-G-CG-G---------G-C--------------------------------------------------C-C---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------TTC-G----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------G--G------------------------------------------GT--C-------AG-C-G-G-C-GG-A--C-------------G-G-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------G--T--G-A--G-T-A---AC-G-C-G--T-G-GG---A-A--C-GT-G--C-C-TTT--CG-G-----------------(deleted rest of line)...
>AP012292.FU3Rum11	97.21	Bacteria;Firmicutes;Negativicutes;Veillonellales-Selenomonadales;Selenomonadaceae;Selenomonas;
```
**Note that:**  

* The fasta file contains dots `.` and dashes `-` 
* The alignment file size if fucking huge because of all that dashes
* The seqID is followed by some other crap in the same line, separated by tabs.

```
$ head -n 5 silva.nr_v138.tax
CP026050.S49M2367	Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Yersiniaceae;Serratia;
JQ864435.O10Acti2	Bacteria;Proteobacteria;Alphaproteobacteria;Rhodobacterales;Rhodobacteraceae;Oceanicella;
AP012292.FU3Rum11	Bacteria;Firmicutes;Negativicutes;Veillonellales-Selenomonadales;Selenomonadaceae;Selenomonas;
KC672487.UncFu112	Eukaryota;Basidiomycota;Tremellomycetes;Filobasidiales;Filobasidiaceae;Heterocephalacria;
JQ771493.AKOEnri2	Archaea;Euryarchaeota;Thermococci;Methanofastidiosales;uncultured;uncultured_ge;
```
**Note that:**  

* The taxonomy file has a tab delimiter btwn seqID and taxonomy names  
* Unclassified nomenclature is all over the place (have to look further to see it all)



## taxonomy formatting  

#### Fill missing names and fix any duplicate names in Silva (R)  

```{bach, eval=F}
$ Rscript force_consistent_nomenclature_on_database.R silva.nr_v138.tax silva.nr_v138_unnamed.tax General
```  
Argument  | Description  
----------|---------------
`force_consistent_nomenclature_on_database.R` | Name of script.  
`silva.nr_v138.tax` | The input taxonomy file, mothur-downloaded.  
`silva.nr_v138_unnamed.tax` | The output taxonomy file, also mothur-formatted. All missing names (blanks or placeholders indicating there is no name at that level) have been changed to `unnamed.NearestCoarserName`. Any duplicate names have had suffixes added to make them unique.  
General | This is the default so this flag could be left out. It tells it to use  `.F` `.G` `.S` instead of `.l` `.c` `.t` fine-level suffixes when it makes duplicated names unique.  

See the FreshTrain call of this script above for more details.  
<br>


## fasta formatting

#### Un-align and delete crap after IDs (R)

This takes ~20 minutes but reduces the file size from ~10GB to MBs. 

```{bash, eval=F}
$ Rscript reformat_fasta.R silva.nr_v138.align silva.nr_v138_unaligned.fasta format.seqIDs
```
Argument  | Description  
----------|---------------
`reformat_fasta.R` | Name of script.
`silva.nr_v138.align` | input mothur-formatted taxonomy file
`silva.nr_v138_unaligned.fasta` | output mothur-formatted taxonomy file with no no U's, dots, or dashes
`format.seqIDs` | This could say anything. As long as the 3rd argument isn't blank, the script will delete any blank space in the sequence ID lines, along with everything after the first blank.
  | Leave 4th argument blank. It would tell it to find all the `U`s and replace them with `T`s but there are no U's in silva so it would just waste time.   









<br>


# Identify Conflicts  

## Classify FreshTrain with Silva  



















# Remove structural conflicts    
A few references in the FreshTrain end up with phylogenetic conflicts (so structural conflicts beyond new names).
These were identified because organisms assigned the same FreshTrain lineage had different SILVA orders, resulting 
in a structure that is not monophyletic. With manual examination, they had very few references in the FreshTrain, so 
to fix this problem we simply remove the less-well-represented half of the references with conflicting names.  
  
  
I started to make a script to do it `remove_non-monophyletic_refs_for_SILVA-compatibility.R` but then I decided
it really wasn't worth the effor because I'm only doing it once. So I manually removed in text wrangler:  
  
### v132 
in FreshTrain these are all "Actinomycetales acSTL."  
in SILVA 132 they are either order "Frankiales" or "PeM15".  

Remove all "PeM15" refs from SILVA v132 FreshTrain-compatible:  
```
AbmAct53  
LaeeEE00p1
PavinActino1
```

### v128

Same issue with acSTL, but different ones. I know, super weird it's not the same.  

Remove all "PeM15" refs from SILVA v128 FreshTrain-compatible:  
```
LakecEC4
LakStec6
PavinActino1
```

Also an issue with betV. In FreshTrain they're all "undefined betV".  
In SILVA 128 they are either order "Nitrosomonadales" or "TRA3-20".  
  
Remove all "TRA3-20" refs from SILVA v128 FreshTrain-compatible:  
```
LiUU2018
LiUU2019
```
